File: code/java/SimpleBenchmark.java
import java.util.Random;

public class SimpleBenchmark {

    public static double[][] randomMatrix(int n, long seed) {
        double[][] M = new double[n][n];
        Random rng = new Random(seed);
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                M[i][j] = rng.nextDouble();
            }
        }
        return M;
    }

    public static double[][] matmulIJK(double[][] A, double[][] B) {
        int n = A.length;
        double[][] C = new double[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                for (int k = 0; k < n; k++) {
                    C[i][j] += A[i][k] * B[k][j];
                }
            }
        }
        return C;
    }

    public static double[][] matmulIKJ(double[][] A, double[][] B) {
        int n = A.length;
        double[][] C = new double[n][n];
        for (int i = 0; i < n; i++) {
            for (int k = 0; k < n; k++) {
                double aik = A[i][k];
                for (int j = 0; j < n; j++) {
                    C[i][j] += aik * B[k][j];
                }
            }
        }
        return C;
    }

    public static void main(String[] args) {
        int[] sizes = {64, 128, 256};
        int runs = 5;

        System.out.println("Java Matrix Multiplication Benchmark");
        System.out.println("=====================================");

        for (int size : sizes) {
            System.out.println("\nMatrix size: " + size + "x" + size);

            double[][] A = randomMatrix(size, 42);
            double[][] B = randomMatrix(size, 43);

            // Benchmark i-j-k
            double[] timesIJK = new double[runs];
            for (int r = 0; r < runs; r++) {
                long start = System.nanoTime();
                matmulIJK(A, B);
                long end = System.nanoTime();
                timesIJK[r] = (end - start) / 1e9;
            }

            // Benchmark i-k-j
            double[] timesIKJ = new double[runs];
            for (int r = 0; r < runs; r++) {
                long start = System.nanoTime();
                matmulIKJ(A, B);
                long end = System.nanoTime();
                timesIKJ[r] = (end - start) / 1e9;
            }

            double meanIJK = mean(timesIJK);
            double meanIKJ = mean(timesIKJ);

            System.out.printf("  i-j-k: %.4fs ± %.4fs%n", meanIJK, stddev(timesIJK));
            System.out.printf("  i-k-j: %.4fs ± %.4fs%n", meanIKJ, stddev(timesIKJ));
            System.out.printf("  Speedup: %.2fx%n", meanIJK / meanIKJ);
        }
    }

    private static double mean(double[] values) {
        double sum = 0;
        for (double v : values) sum += v;
        return sum / values.length;
    }

    private static double stddev(double[] values) {
        double m = mean(values);
        double sum = 0;
        for (double v : values) sum += (v - m) * (v - m);
        return Math.sqrt(sum / values.length);
    }
}

File: code/cpp/simple_benchmark.cpp
#include "include/matrix_mul.h"
#include <chrono>
#include <iostream>
#include <numeric>
#include <iomanip>

double calculate_mean(const std::vector<double>& times) {
    return std::accumulate(times.begin(), times.end(), 0.0) / times.size();
}

double calculate_stddev(const std::vector<double>& times) {
    double mean = calculate_mean(times);
    double sum = 0.0;
    for (double time : times) {
        sum += (time - mean) * (time - mean);
    }
    return std::sqrt(sum / times.size());
}

int main() {
    std::vector<int> sizes = {64, 128, 256};
    int runs = 5;

    std::cout << "C++ Matrix Multiplication Benchmark" << std::endl;
    std::cout << "====================================" << std::endl;

    for (int size : sizes) {
        std::cout << "\nMatrix size: " << size << "x" << size << std::endl;

        auto A = random_matrix(size, 42);
        auto B = random_matrix(size, 43);

        // Benchmark i-j-k
        std::vector<double> times_ijk(runs);
        for (int r = 0; r < runs; r++) {
            auto start = std::chrono::high_resolution_clock::now();
            auto C = matmul_ijk(A, B);
            auto end = std::chrono::high_resolution_clock::now();
            times_ijk[r] = std::chrono::duration<double>(end - start).count();
        }

        // Benchmark i-k-j
        std::vector<double> times_ikj(runs);
        for (int r = 0; r < runs; r++) {
            auto start = std::chrono::high_resolution_clock::now();
            auto C = matmul_ikj(A, B);
            auto end = std::chrono::high_resolution_clock::now();
            times_ikj[r] = std::chrono::duration<double>(end - start).count();
        }

        double mean_ijk = calculate_mean(times_ijk);
        double mean_ikj = calculate_mean(times_ikj);
        double std_ijk = calculate_stddev(times_ijk);
        double std_ikj = calculate_stddev(times_ikj);

        std::cout << std::fixed << std::setprecision(4);
        std::cout << "  i-j-k: " << mean_ijk << "s ± " << std_ijk << "s" << std::endl;
        std::cout << "  i-k-j: " << mean_ikj << "s ± " << std_ikj << "s" << std::endl;
        std::cout << "  Speedup: " << std::setprecision(2) << mean_ijk/mean_ikj << "x" << std::endl;
    }

    return 0;
}